File: src/bootcamp/hello-sim-object/HelloSimObject.py
# 导入gem5所有模块的基类：SimObject（所有自定义硬件模块必须继承它）
from m5.objects.SimObject import SimObject
# 导入gem5的参数定义工具（Param是核心，用来给模块加配置）
from m5.params import *

# 第一个自定义模块：HelloSimObject
class HelloSimObject(SimObject):
    # 1. gem5内部唯一识别名（必须和C++类名一致）
    type = "HelloSimObject"
    # 2. 绑定：这个Python类对应的C++头文件路径
    cxx_header = "bootcamp/hello-sim-object/hello_sim_object.hh"
    # 3. 绑定：这个Python类对应的C++类全名（带命名空间）→ C++gem5::HelloSimObject
    cxx_class = "gem5::HelloSimObject"
    # 4. 定义整数参数 → 你最困惑的点2 参数名：num_hellos  注释"Number of times to say Hello."  类型： int
    num_hellos = Param.Int("Number of times to say Hello.")
    # 5. 定义模块关联参数 → 你最困惑的点3
    goodbye_object = Param.GoodByeSimObject("GoodByeSimObject to say goodbye after done saying hello.")

# 第二个自定义模块：GoodByeSimObject
class GoodByeSimObject(SimObject):
    type = "GoodByeSimObject"
    cxx_header = "bootcamp/hello-sim-object/goodbye_sim_object.hh"
    cxx_class = "gem5::GoodByeSimObject"

File: src/bootcamp/hello-sim-object/hello_sim_object.cc
#include "bootcamp/hello-sim-object/hello_sim_object.hh"

#include <iostream>

#include "base/trace.hh"
#include "debug/HelloExampleFlag.hh"
#include "sim/eventq.hh"

namespace gem5
{
    // 构造函数：接收Python传递的所有参数（params包含num_hellos、goodbye_object）
    HelloSimObject::HelloSimObject(const HelloSimObjectParams& params):
    // 1. 初始化父类 SimObject（gem5 硬性要求）
    SimObject(params),
    // 2. 初始化剩余打印次数 = Python传入的 num_hellos（比如5）
    remainingHellosToPrintByEvent(params.num_hellos),
    // 3. 初始化事件对象：绑定要执行的函数 + 给事件命名
    nextHelloEvent([this](){ processNextHelloEvent(); }, name() + "nextHelloEvent"),
    // 4. 初始化指针：绑定Python传入的 GoodByeSimObject 对象
    goodByeObject(params.goodbye_object)
    {
        // 校验参数：如果num_hellos≤0，直接报错退出
        fatal_if(params.num_hellos <= 0, "num_hellos should be positive!");
        // 循环打印：构造函数里直接输出 Hello（无延时，一次性打完）
        for (int i = 0; i < params.num_hellos; i++) {
            std::cout << "i: " << i << ", Hello from HelloSimObject's constructor!" << std::endl;
        }
        // gem5 调试日志（只有开启调试开关时才打印）
        DPRINTF(HelloExampleFlag, "%s: Hello from HelloSimObject's constructor!\n", __func__);
    }

    void
    HelloSimObject::processNextHelloEvent()
    {
        std::cout << "tick: " << curTick() << ", Hello from HelloSimObject::processNextHelloEvent!" << std::endl;
        remainingHellosToPrintByEvent--;
        if (remainingHellosToPrintByEvent > 0) {
            schedule(nextHelloEvent, curTick() + 500);
        } 
        else {
            goodByeObject->sayGoodBye();
        }
    }

    void
    HelloSimObject::startup()
    {
        std::cout << "START_UP_HELLO" << std::endl;
        panic_if(curTick() != 0, "startup called at a tick other than 0");
        panic_if(nextHelloEvent.scheduled(), "nextHelloEvent is scheduled before HelloSimObject::startup is called!");
        schedule(nextHelloEvent, curTick() + 500);
    }
} // namespace gem5
